Propagate -DPCRE_STATIC from pcre.BUILD to swig.BUILD

snnn opened the pull request:
pcre_internal.h uses this macro to determine the content of PCRE_EXP_DECL. So, in order to keep consistent, every file who includes pcre_internal.h  directly or indirectly should also have this macro defined.

This pull request is for fix a build error on Windows:
<hr/>
ERROR: C:/os/t/external/swig/BUILD.bazel:5:1: Linking of rule '@swig//:swig' failed (Exit 1120): link.exe failed: error executing command                          <br/>
misc.o : error LNK2019: unresolved external symbol __imp_pcre_compile referenced in function Swig_string_regex <br/>                      
naming.o : error LNK2001: unresolved external symbol __imp_pcre_compile                       
misc.o : error LNK2019: unresolved external symbol __imp_pcre_exec referenced in function Swig_string_regex   <br/>                    
naming.o : error LNK2001: unresolved external symbol __imp_pcre_exec        <br/>               
misc.o : error LNK2019: unresolved external symbol __imp_pcre_version referenced in function Swig_pcre_version                       <br/>
misc.o : error LNK2001: unresolved external symbol __imp_pcre_free            <br/>           
naming.o : error LNK2001: unresolved external symbol __imp_pcre_free           <br/>            
bazel-out/host/bin/external/swig/swig.exe : fatal error LNK1120: 4 unresolved externals    <br/>                   
<hr/>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

vrv commented:
Ping for @jart.

I'm a little worried about this because using 'defines' propagates these defines to all libraries that depend on on this.  Given that these defines are very generic sounding, it's likely to cause conflicts in the future.  I don't have a good suggestion on how to deal with this though.

snnn commented:
Hi @vrv Currently, pcre is private to swig, not visible to the others.

jart commented:
If `PCRE_STATIC` is the only one that needs to be propagated, then why not have that be the only one listed under `defines` and have the rest remain copts?

snnn commented:
Hi @jart  If you search “HAVE_” in pcre_internal.h, there are a lot of occurrences.

jart commented:
@snnn The public headers don't appear to include the internal header. Only the .c. files do. So those defines should never be needed by downstream rules. Therefore it should be sufficient to say:

```py
    defs = ["PCRE_STATIC=1"],
    copts = [
        "-DHAVE_BCOPY=1",
        "-DHAVE_INTTYPES_H=1",
        "-DHAVE_MEMMOVE=1",
        "-DHAVE_STDINT_H=1",
        "-DHAVE_STRERROR=1",
        "-DHAVE_SYS_STAT_H=1",
        "-DHAVE_SYS_TYPES_H=1",
        "-DHAVE_UNISTD_H=1",
        "-DLINK_SIZE=2",
        "-DMATCH_LIMIT=10000000",
        "-DMATCH_LIMIT_RECURSION=1000",
        "-DMAX_NAME_COUNT=10000",
        "-DMAX_NAME_SIZE=32",
        "-DNEWLINE=10",
        "-DNO_RECURSE",
        "-DPARENS_NEST_LIMIT=50",
        "-DPOSIX_MALLOC_THRESHOLD=10",
        "-DSTDC_HEADERS=1",
        "-DSUPPORT_UCP",
        "-DSUPPORT_UTF",
    ],
```

You're correct in what you said earlier, that this wouldn't be an issue due to the visibility restriction. To offer more context on why this is worth doing, simply for its own sake, consider https://github.com/tensorflow/tensorflow/commit/d88cccebc7f61078d775d26f4714a06bc4002fcf. We actually got in trouble a few days ago because we were using a overly broad viral define. It ended up colliding with the codebase of a random team at Google.

Now that you know the reasoning, I want to thank you again for your contribution, and hope you'll help us set a positive example for anyone reading this build config in the future.

snnn commented:
Hi @jart 

Thanks for pointing it out.  I'll update this PR according to your suggestion.

Commit:
Propagate -DPCRE_STATIC from pcre.BUILD to swig.BUILD

To fix a build error on Windows:
ERROR: C:/os/t/external/swig/BUILD.bazel:5:1: Linking of rule '@swig//:swig'
  failed (Exit 1120): link.exe failed: error executing command
misc.o : error LNK2019: unresolved external symbol __imp_pcre_compile
  referenced in function Swig_string_regex
...

jart commented:
Thank you. Please ping when it's ready for approval.

snnn commented:
Let's hold on this PR until https://github.com/bazelbuild/bazel/issues/3949 is fixed. Otherwise, I'm not able to test it.

jart reviewed (approved):
I'd actually like to just go ahead and submit anyway, because I don't see the harm.

snnn commented:
Please do it.

jart commented:
Jenkins test this please.

gunan commented:
Jenkins, test this please.
